Update session definition with transducer tracking result

sadhana-r opened the pull request:
Closes #182

This PR should be merged after #188.

ebrahimebrahim commented on src/openlifu/db/session.py:
```diff
@@ -72,9 +92,8 @@ class Session:
-    transducer_tracking_approved: Optional[bool] = False
-    """Approval state of transducer tracking. `True` means the user has provided some kind of
-    confirmation that the transducer transform in this session agrees with reality."""
+    transducer_tracking_results: Optional[List[TransducerTrackingResult]] = field(default_factory=list)
+    """List of any transducer tracking results"""
```
It can just be a List instead of an Optional[List]

(the case of an empty list is meaningful, and I can't think of anything else that `None` could mean)

ebrahimebrahim commented on src/openlifu/db/session.py:
```diff
@@ -121,6 +140,11 @@ def from_dict(d:Dict):
             raise ValueError("Sessions no longer recognize a volume attribute -- it is now volume_id.")
         if 'array_transform' in d:
             d['array_transform'] = ArrayTransform(np.array(d['array_transform']['matrix']), d['array_transform']['units'])
+        if 'transducer_tracking_results' in d:
+            d['transducer_tracking_results'] = [TransducerTrackingResult(t['photoscan_id'],
+                                                                         ArrayTransform(np.array(t['transducer_to_photoscan_transform']['matrix']),t['transducer_to_photoscan_transform']['units']),
+                                                                         ArrayTransform(np.array(t['photoscan_to_volume_transform']['matrix']), t['photoscan_to_volume_transform']['units']),
+                                                                         t['transducer_tracking_approved']) for t in d['transducer_tracking_results']]
```
More readable indentation:

```suggestion
        if 'transducer_tracking_results' in d:
            d['transducer_tracking_results'] = [
                TransducerTrackingResult(
                    t['photoscan_id'],
                    ArrayTransform(np.array(t['transducer_to_photoscan_transform']['matrix']),t['transducer_to_photoscan_transform']['units']),
                    ArrayTransform(np.array(t['photoscan_to_volume_transform']['matrix']), t['photoscan_to_volume_transform']['units']),
                    t['transducer_tracking_approved']
                )
                for t in d['transducer_tracking_results']
            ]
```

ebrahimebrahim commented on tests/resources/example_db/subjects/example_subject/sessions/example_session/example_session.json:
```diff
@@ -28,5 +28,28 @@
+      "photoscan_id": "example_photoscan",
+      "transducer_to_photoscan_transform": {
+        "matrix": [
+          [1.0, 0.0, 0.0, 0.0],
+          [0.0, 1.0, 0.0, 0.0],
+          [0.0, 0.0, 1.0, 0.0],
+          [0.0, 0.0, 0.0, 1.0]
+        ],
+        "units": "mm"
+      },
+      "photoscan_to_volume_transform": {
+        "matrix": [
+          [1.0, 0.0, 0.0, 0.0],
+          [0.0, 1.0, 0.0, 0.0],
+          [0.0, 0.0, 1.0, 0.0],
+          [0.0, 0.0, 0.0, 1.0]
```
Could make these matrices slightly different from the  identity and different from each other, just to help make the tests slightly pickier

ebrahimebrahim commented on db_dvc.dvc:
```diff
@@ -1,6 +1,6 @@
-- md5: 964978fc7c5a135c71de8e2ea547e0e4.dir
-  size: 802948419
-  nfiles: 68
+- md5: 83e1023daf8e32799d4c9f308fd3f711.dir
```
I'll check the DVC data once that is rebased

ebrahimebrahim reviewed (commented):
Started reviewing, but then realized this might need to be rebased? I'll come back and finish the review after that. For now it pretty much looks good, thanks!

ebrahimebrahim commented on src/openlifu/db/database.py:
```diff
@@ -192,7 +201,7 @@ def write_subject(self, subject, on_conflict=OnConflictOpts.ERROR):
         self.logger.info(f"Added subject with ID {subject_id} to the database.")
 
-    def write_transducer(self, transducer, on_conflict: OnConflictOpts=OnConflictOpts.ERROR):
+    def write_transducer(self, transducer, registration_surface_model_filepath: Optional[str] = None, transducer_body_model_filepath: Optional[str] = None, on_conflict: OnConflictOpts=OnConflictOpts.ERROR):
```
I was getting deja vu looking at this line....

It looks like this already got into main. So I guess this was branched off of somehting that already got merged?

In that case you can do an interactive rebase and remove the commits that already got into main. LMK if you want any help with that, I have some idea of how to go about it if it ends up being not easy. Thanks!

sadhana-r commented:
Whoops, forgot this was very out of date now. Rebased to main, updated the dvc data and addressed your comments. 

As a reminder for us: **Do not merge** until SlicerOpenLIFU PR addressing : https://github.com/OpenwaterHealth/SlicerOpenLIFU/issues/157 is merged. The changes mades to sessions in this PR break SliceOpenLIFU.

ebrahimebrahim commented on tests/test_database.py:
```diff
@@ -290,6 +298,21 @@ def test_write_session_associated_object_structure_created(example_database: Dat
     assert example_database.get_runs_filename(example_subject.id, session.id).is_file()
 
+def test_write_session_with_invalid_photoscan_id(example_database: Database, example_subject: Subject, example_transducer_tracking_result):
+    """ Test that when there is a transducer tracking result class associated with a session, the session
+    is correctly written to file."""
```
Looks like this docstring got swapped with the docstring on the test just below it

ebrahimebrahim reviewed (commented):
Small comment, otherwise it looks good! I was unable to check the DVC data though, I think maybe you forgot to `dvc push`?

ebrahimebrahim reviewed (approved):
Thank you!

> **Do not merge** until SlicerOpenLIFU PR addressing : [OpenwaterHealth/SlicerOpenLIFU#157](https://github.com/OpenwaterHealth/SlicerOpenLIFU/issues/157) is merged. The changes mades to sessions in this PR break SliceOpenLIFU.

sadhana-r commented:
@ebrahimebrahim could you check the db_dvc for this PR once more? I moved the transducer tracking result to  the `demo_4_simulated` session - before it was incorrectly in `example_session` which doesn't have an affiliated photoscan. I also made sure that every session now has a transducer_tracking_result attribute, initialized with `[]`. Notably, I have added a new session under mannequin called `demo_4_tt` that has a decimated photoscan from the registration example data that we can use for demoing instead of the trumpet, and the session uses transducer `curved2_200roc_10gap` which has the registration surface needed for transducer tracking. 
I have also now updated the transducer tracking result class to include separate approval flags for each transform.

ebrahimebrahim commented:
The DVC data looks good to me.
Comparing against the current main branch, I see exactly the changes you listed and nothing more.
(I didn't test that everything loads in SlicerOpenLIFU)

This all looks great!!

---

Just because I got burned by it before, I recommend adding a line to here to test that when you read a session the things that are supposed to be numpy arrays really are actually read as numpy arrays:

https://github.com/OpenwaterHealth/OpenLIFU-python/blob/322c75a3ea7b3eeba93aeb4bea7ab092412cf4a6/tests/test_database.py#L399-L404

You can just add a bit to the end of that test, something like

```py
for tt_result in example_session.transducer_tracking_results:
    assert isinstance(tt_result.transducer_to_photoscan_transform.matrix, np.ndarray)
    assert isinstance(tt_result.photoscan_to_volume_transform.matrix, np.ndarray)
```
